mod/quiz/accessrule/seb/classes/access_manager.php

```diff
@@ -73,48 +73,105 @@ public function __construct(quiz $quiz) {
     }
 
     /**
-     * Check if the browser exam key hash in header matches one of the listed browser exam keys from quiz settings.
+     * Validate SEB keys that should be checked. If the keys are valid, store the access state in Moodle session.
+     *
+     * @param string|null $configkey Optionally provide a SEB config key to validate.
+     * @param string|null $browserexamkey Optionally provide a SEB browser exam key to validate.
+     * @param string|null $url Optionally provide URL of page to validate.
+     * @return bool
+     */
+    public function validate_seb_keys(?string $configkey = null, ?string $browserexamkey = null, ?string $url = null): bool {
+        // Check if there is a valid config key.
+        if ($this->should_validate_config_key() && !$this->validate_config_key($configkey, $url)) {
+            return false;
+        }
+
+        // Check if there is a valid browser exam key.
+        if ($this->should_validate_browser_exam_key() && !$this->validate_browser_exam_key($browserexamkey, $url)) {
+            return false;
+        }
+
+        // Set the state of the access for this Moodle session.
+        $this->set_session_access(true);
+
+        return true;
+    }
+
+    /**
+     * Validate browser exam key. It will validate a provided browser exam key if provided, then will fall back to checking
+     * the header.
      *
-     * @return bool True if header key matches one of the saved keys.
+     * @param string|null $browserexamkey Optional. Can validate a provided key, or will fall back to checking header.
+     * @return bool
      */
-    public function validate_browser_exam_keys() : bool {
-        // If browser exam keys are entered in settings, check they match the header.
-        $browserexamkeys = $this->quizsettings->get('allowedbrowserexamkeys');
-        if (empty($browserexamkeys)) {
+    public function validate_browser_exam_key(?string $browserexamkey = null, ?string $url = null): bool {
+        if (!$this->should_validate_browser_exam_key()) {
+            throw new \coding_exception('Check whether the browser exam key should be validated with $accessmanager->should_validate_browser_exam_key()');
```

Comment:
I'm a little sceptical about this exception for that public method. It feels wrong to throw it here. But I get why you do that. 
I think technically we can return true here is it shouldn't be validated at all. Means we will not prevent access and basically won't validate it at all.

Another option could be to make this method private, then it's 100% legit to throw that exception and expect this method to be called ONLY after checking for should_validate_browser_exam_key.

Thoughts?

Reply:
Yeah, you're right. I had to change my thinking from deny unless valid to allow unless invalid. Have replaced coding exceptions with `return true;`. 

This means that we don't need to call the 'should' functions before calling the validate functions that makes the code a little more robust.

Can't make them private, regardless, as have removed the 'validate_seb_keys' refactor so they are called directly.